### PR TITLE
Add Docker Desktop Edge 2.3.1.0.45408

### PR DESCRIPTION
**Pull request by SoftCreatR**
Auto Generated by softcreatr@winget.it on Thu, 28 May 2020 10:05:38 +0000

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/winget-pkgs/pull/1275)

### CODE REVIEW

**Comment by suneetk92** [comment]
tag is missing
```Channel: edge```

**Comment by suneetk92** [comment]
no end of line present.
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

**Comment by SoftCreatR** [comment]
> no end of line present.
> https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

Doesn't matter. But the next releases will have it ;)



> tag is missing
> `Channel: edge`

See https://github.com/microsoft/winget-pkgs/pull/1224#issuecomment-635950452

**Review by ghost — APPROVED** [review]
Validation has completed

**Comment by StefanScherer** [comment]
@KevinLaMS Thank you that you are working so hard catching up all pull requests. I work for Docker Inc. and want to give our users a good experience getting Docker Desktop. Winget is an interesting tool for developers to automate their machine setups.

After this PR got merged, I'm not really happy with the current situation. Now with two packages it got more complicated and confusion for users to install Docker Desktop via `winget`.

```
PS C:\Users\stefan\Desktop> winget search docker
Name                Id                       Version       Matched
--------------------------------------------------------------------------
Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519 Moniker: docker
Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408 Moniker: docker


PS C:\Users\stefan\Desktop> winget install docker
Multiple apps found matching input criteria. Please refine the input.
Name                Id                       Version       Matched
--------------------------------------------------------------------------
Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519 Moniker: docker
Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408 Moniker: docker


PS C:\Users\stefan\Desktop> winget install dockerdesktop
Multiple apps found matching input criteria. Please refine the input.
Name                Id                       Version
----------------------------------------------------------
Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519
Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408


PS C:\Users\stefan\Desktop> winget install docker.dockerdesktop
Multiple apps found matching input criteria. Please refine the input.
Name                Id                       Version
----------------------------------------------------------
Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519
Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408


PS C:\Users\stefan\Desktop> winget install docker.dockerdesktopedge
Found Docker Desktop Edge [Docker.DockerDesktopEdge]
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://desktop.docker.com/win/edge/45408/Docker%20Desktop%20Installer.exe
...
```

What can we do to improve the user experience to get the Stable release for `winget install docker` (which worked until the merge) or `winget install dockerdesktop`?

Even a `winget install  docker.dockerdesktop` does not work any longer.

When I compare that to chocolatey they provide the user the option like this
`choco install docker-desktop` for Stable and `choco install -pre docker-desktop` for Edge version.

File: manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# ================== General ==================
+
+Id: Docker.DockerDesktopEdge
+Publisher: Docker Inc.
+Name: Docker Desktop Edge
+AppMoniker: docker
```
Comment (StefanScherer) [inline manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml:6]: This should be removed I guess, we now have two monikers.

**Comment by SoftCreatR** [comment]
@StefanScherer https://github.com/microsoft/winget-pkgs/pull/1387

For the naming/moniker, i've already created a patch for our internal updater, so the next release of Docker Desktop Edge will have the Moniker "docker-edge", while Docker Desktop remains "docker".

**Comment by suneetk92** [comment]
T

> 
> 
> @KevinLaMS Thank you that you are working so hard catching up all pull requests. I work for Docker Inc. and want to give our users a good experience getting Docker Desktop. Winget is an interesting tool for developers to automate their machine setups.
> 
> After this PR got merged, I'm not really happy with the current situation. Now with two packages it got more complicated and confusion for users to install Docker Desktop via `winget`.
> 
> ```
> PS C:\Users\stefan\Desktop> winget search docker
> Name                Id                       Version       Matched
> --------------------------------------------------------------------------
> Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519 Moniker: docker
> Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408 Moniker: docker
> 
> 
> PS C:\Users\stefan\Desktop> winget install docker
> Multiple apps found matching input criteria. Please refine the input.
> Name                Id                       Version       Matched
> --------------------------------------------------------------------------
> Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519 Moniker: docker
> Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408 Moniker: docker
> 
> 
> PS C:\Users\stefan\Desktop> winget install dockerdesktop
> Multiple apps found matching input criteria. Please refine the input.
> Name                Id                       Version
> ----------------------------------------------------------
> Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519
> Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408
> 
> 
> PS C:\Users\stefan\Desktop> winget install docker.dockerdesktop
> Multiple apps found matching input criteria. Please refine the input.
> Name                Id                       Version
> ----------------------------------------------------------
> Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519
> Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408
> 
> 
> PS C:\Users\stefan\Desktop> winget install docker.dockerdesktopedge
> Found Docker Desktop Edge [Docker.DockerDesktopEdge]
> This application is licensed to you by its owner.
> Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
> Downloading https://desktop.docker.com/win/edge/45408/Docker%20Desktop%20Installer.exe
> ...
> ```
> 
> What can we do to improve the user experience to get the Stable release for `winget install docker` (which worked until the merge) or `winget install dockerdesktop`?
> 
> Even a `winget install docker.dockerdesktop` does not work any longer.
> 
> When I compare that to chocolatey they provide the user the option like this
> `choco install docker-desktop` for Stable and `choco install -pre docker-desktop` for Edge version.

Try winget install --id Docker.DockerDesktop -e

File: manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Description: Docker Desktop is an application for MacOS and Windows machines for the building and sharing of containerized applications. Access Docker Desktop and follow the guided onboarding to build your first containerized application in minutes.
+Homepage: https://www.docker.com
+Tags: Container, Containerization, Virtualization
+
+Version: 2.3.1.0.45408
```
Comment (StefanScherer) [inline manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml:11]: Why 5 numbers?
According to https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv0.1.md there should be only 4, the installer version also does not contain our internal build number. It's part of the download URL, but shouldn't appear here as a `winget search` then looks more complicated

```
PS C:\Users\stefan\Desktop> winget search docker
Name                Id                       Version       Matched
--------------------------------------------------------------------------
Docker Desktop      Docker.DockerDesktop     2.3.0.3.45519 Moniker: docker
Docker Desktop Edge Docker.DockerDesktopEdge 2.3.1.0.45408 Moniker: docker
```

Comment (SoftCreatR) [reply manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml:11]: > does not contain our internal build number. It's part of the download URL

You sure? https://download.docker.com/win/stable/appcast.xml

> According to https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv0.1.md there should be only 4

As of now, you can ignore the specification.

File: manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml (hunk 1 above)
Comment (SoftCreatR) [reply manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml:6]: https://github.com/microsoft/winget-pkgs/pull/1390

File: manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml (hunk 2 above)
Comment (StefanScherer) [reply manifests/Docker/DockerDesktopEdge/2.3.1.0.45408.yaml:11]: Alright, so let's keep it as it is right now. Thanks for the details.

**Comment by StefanScherer** [comment]
Thanks @SoftCreatR and @suneetk92 indeed `winget install --id Docker.DockerDesktop -e` works, but to be honest this is not really intuitive 😅 
Looking forward to the fixed monikers 👍

**Comment by SoftCreatR** [comment]
@Stefanscherer I am not 100% confident, that it works as planned, because as long as two versions of a package exist and as long as winget doesn't support channels, it might still return two versions, because both start with Docker.

**Comment by StefanScherer** [comment]
Is there a way to test this locally without merging PR's "to production"? :-)

**Comment by SoftCreatR** [comment]
In theory, this should be somehow possible using the [source](https://docs.microsoft.com/de-de/windows/package-manager/winget/source) command. Unfortunately, i don't understand, how this actually works 😄

**Comment by suneetk92** [comment]
> 
> 
> Is there a way to test this locally without merging PR's "to production"? :-)

Did you mean testing yaml file locally?

you can do by executing 
`winget validate <.yaml>`
or
`winget install -m <.yaml>`

> 
> 
> In theory, this should be somehow possible using the [source](https://docs.microsoft.com/de-de/windows/package-manager/winget/source) command. Unfortunately, i don't understand, how this actually works 😄

They work similar to helm charts' repo.
You can add your personal / private repository by using
`winget source add test https://winget.azureedge.net/cache`

**Comment by SoftCreatR** [comment]
> Did you mean testing yaml file locally?
> 
> you can do by executing
> `winget validate <.yaml>`
> or
> `winget install -m <.yaml>`

Nah, he's talking about the package search with the modified AppMoniker.

**Pull request merged**